src/functional/otf/stages.py

```diff
@@ -0,0 +1,116 @@
+# GT4Py Project - GridTools Framework
+#
+# Copyright (c) 2014-2021, ETH Zurich
+# All rights reserved.
+#
+# This file is part of the GT4Py project and the GridTools framework.
+# GT4Py is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+from __future__ import annotations
+
+import dataclasses
+from typing import Any, Generic, Optional, Protocol, TypeVar
+
+from functional.iterator import ir as itir
+from functional.otf import languages
+from functional.otf.source import source
+
+
+SrcL = TypeVar("SrcL", bound=languages.LanguageTag)
+TgtL = TypeVar("TgtL", bound=languages.LanguageTag)
+SettingT = TypeVar("SettingT", bound=languages.LanguageSettings)
+SrcL_co = TypeVar("SrcL_co", bound=languages.LanguageTag, covariant=True)
+TgtL_co = TypeVar("TgtL_co", bound=languages.LanguageTag, covariant=True)
+SettingT_co = TypeVar("SettingT_co", bound=languages.LanguageSettings, covariant=True)
+
+
+@dataclasses.dataclass(frozen=True)
+class ProgramCall:
+    """Iterator IR representaion of a program together with arguments to be passed to it."""
+
+    program: itir.FencilDefinition
+    args: tuple[Any, ...]
+    kwargs: dict[str, Any]
+
+
+@dataclasses.dataclass(frozen=True)
+class ProgramSource(Generic[SrcL, SettingT]):
+    """
+    Standalone sourcecode translated from an IR representation along with OTF relevant information.
```

Comment:
Probably best to avoid nested multi-part adjectives like *(on-the-fly)-compilation-relevant*:
```suggestion
    Standalone source code translated from an IR along with information relevant for OTF compilation.
```